gdal/frmts/sentinel2/sentinel2dataset.cpp

```diff
@@ -881,17 +928,36 @@ static bool SENTINEL2GetGranuleList(CPLXMLNode* psMainMTD,
 
             /* S2A_OPER_MSI_L1C_TL_SGS__20151024T023555_A001758_T53JLJ_N01.04 --> */
             /* S2A_OPER_MTD_L1C_TL_SGS__20151024T023555_A001758_T53JLJ */
+            // S2B_OPER_MSI_L2A_TL_MPS__20180823T122014_A007641_T34VFJ_N02.08
             CPLString osGranuleMTD = pszGranuleId;
             if( osGranuleMTD.size() > strlen("S2A_OPER_MSI_") &&
                 osGranuleMTD[8] == '_' && osGranuleMTD[12] == '_' &&
                 osGranuleMTD[osGranuleMTD.size()-7] == '_' &&
-                osGranuleMTD[osGranuleMTD.size()-6] == 'N' )
+                osGranuleMTD[osGranuleMTD.size()-6] == 'N' &&
+                osGranuleMTD[5] == 'O' )
             {
                 osGranuleMTD[9] = 'M';
                 osGranuleMTD[10] = 'T';
                 osGranuleMTD[11] = 'D';
                 osGranuleMTD.resize(osGranuleMTD.size()-7);
             }
+            if( isMSI2Ap )
+            {
+                osGranuleMTD = "MTD_TL";
+                oGranuleId = "L2A_";
+                // S2A_MSIL2A_20170823T094031_N0205_R036_T34VFJ_20170823T094252.SAFE
+                // S2A_USER_MSI_L2A_TL_SGS__20170823T133142_A011330_T34VFJ_N02.05 -->
+                // L2A_T34VFJ_A011330_20170823T094252
+                const char* pszProductURI = CPLGetXMLValue(psProductInfo, "PRODUCT_URI_2A", nullptr);
+                if( pszProductURI != nullptr )
+                {
+                    CPLString vProductURI(pszProductURI);
```

Comment:
there should be a safety check on vProductURI so that below string manipulations do not go out-of-range for corrupted/hostile datasets

Reply:
Size checks are implemented.